### PR TITLE
Remove the `NEWOBJ` and `OBJSETUP` macros

### LINKED ISSUE #2869: Add `rb_newobj` to native extension API
_memcached_ 1.8.0 requires the `rb_newobj` function via the `NEWOBJ` macro. #2752 implemented several new extension functions for _memcached_, but at the time I narrowed the scope down to just those used by a particular application I was trying to get running. I'm now adding some of that application's dependencies to CI with TruffleRuby and finding more functionality being exercised. It's probably worth just running the _memcached_ test suite and see what else we're missing.

### PR DESCRIPTION
**Pull request by nirvdrum**
Remove the `NEWOBJ` and `OBJSETUP` macros since we cannot support them in TruffleRuby and native extensions may use `#ifdef` to detect features.

Notably, this fixes compatibility with _memcached_ 1.8.0. Fixes #2869.

### CODE REVIEW

**Comment by nirvdrum** [comment]
The test failures appear to be related to the TTY bug in the latest JDK we're using. The fast specs passed locally. I can run the others manually as well.

File: lib/cext/include/ruby/internal/newobj.h [hunk 1]
```diff
@@ -57,9 +57,12 @@
  */
 #define RB_NEWOBJ_OF(obj,type,klass,flags) type *(obj) = RBIMPL_CAST((type *)rb_newobj_of(klass, flags))
 
-#define NEWOBJ     RB_NEWOBJ      /**< @old{RB_NEWOBJ} */
+// We can't support either `NEWOBJ` or `OBJSETUP` in TruffleRuby. They rely on being able to create an object
+// without an associated class and updating state that is tied directly to MRI's object layout.
+// #define NEWOBJ     RB_NEWOBJ      /**< @old{RB_NEWOBJ} */
+// #define OBJSETUP   rb_obj_setup   /**< @old{rb_obj_setup} */
```
Comment (eregon) [inline lib/cext/include/ruby/internal/newobj.h:63]: So in headers we import from CRuby we never change lines but only add lines (which is better for conflicts and clarity and seeing the differences), so this should be a `#ifndef TRUFFLERUBY` around these.

**Comment by eregon** [comment]
It would be great to add a spec in `spec/truffle/capi` that those 2 macros are not defined, so we don't regress accidentally, e.g. when importing 3.2 headers (I typically try to remove extra ifdefs when importing new headers, here the comment would save it probably but still, the macro might be defined somewhere else, etc).

File: lib/cext/include/ruby/internal/newobj.h (hunk 1 above)
Comment (nirvdrum) [reply lib/cext/include/ruby/internal/newobj.h:63]: Fair enough. I just wanted the comment grouped with the lines changed, but I can leave it floating above.

**Comment by nirvdrum** [comment]
@eregon I've added some specs. I think this addresses what you're looking for. I don't love the all caps parts of the method names, but I do like that it's clear we're referring to the `#define` values.

I suppose that leads to another point... this PR removes the macros but retains `rb_newobj` and `rb_obj_setup`. I don't think we want to remove function declarations, but they naturally won't work either. Would it make sense to implement them and raise an exception message about them being unsupported? Right now we end up with something like:

```
Polyglot::ForeignException:
    External LLVMFunction rb_newobj cannot be found.
```

That message doesn't mean much to someone not working on TruffleRuby. Maybe it'd help guide extension authors to a better choice rather than having them think TruffleRuby is broken. Maybe we could link back to the docs on why a particular function isn't implemented and suggested fixes. That's probably out of scope here, though.

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
Regarding these functions, I think we should change nothing to them. The docs say:
```
/**
 * This is the implementation detail of #RB_NEWOBJ.
 *
 * @exception  rb_eNoMemError  No space left.
 * @return     An allocated object, not initialised.
 */
VALUE rb_newobj(void);
```
So it's clear, it's incorrect to use those in C extensions.

And we cannot have a custom error message for all of these (way too much work).
BTW, they should also fail at link time if used, which is the proper time to fail for such things.

**Comment by eregon** [comment]
What we should do though is make sure `RB_NEWOBJ` is not defined either, I'll do that.

**Comment by eregon** [comment]
> BTW, they should also fail at link time if used, which is the proper time to fail for such things.

I found the issue I was thinking of, it's https://bugs.ruby-lang.org/issues/18615.
Which we already have, but for that to trigger we need to actually remove the function prototypes.
But because people shouldn't use these functions directly it shouldn't matter.

**Comment by eregon** [comment]
FYI I was wondering why the ABI check didn't fail on GitHub, but failed in the internal gate.
It's because GitHub does a strange git clone where there is a single commit with the same hash as the last commit of this PR, and there is therefore no parent commit and so that only commit is considered to have updated the ABI version file even though it did not.

**Pull request merged**